servicetalk-concurrent-api/src/main/java/io/servicetalk/concurrent/api/Publisher.java

```diff
@@ -2256,23 +2310,83 @@ public final Publisher<T> retry(BiIntPredicate<Throwable> shouldRetry) {
-     *
      * @see <a href="https://reactivex.io/documentation/operators/retry.html">ReactiveX retry operator.</a>
+     * @see #retryWhen(boolean, BiIntFunction)
      */
     public final Publisher<T> retryWhen(BiIntFunction<Throwable, ? extends Completable> retryWhen) {
-        return new RedoWhenPublisher<>(this, (retryCount, notification) -> {
-            if (notification.cause() == null) {
-                return completed();
-            }
-            return retryWhen.apply(retryCount, notification.cause());
-        }, true);
+        return retryWhen(false, retryWhen);
+    }
+
+    /**
+     * Re-subscribes to this {@link Publisher} if an error is emitted and the {@link Completable} returned by the
+     * supplied {@link BiIntFunction} completes successfully. If the returned {@link Completable} emits an error, the
+     * returned {@link Publisher} terminates with that error.
+     * <pre>
+     * This method may result in a {@link StackOverflowError} if too many consecutive calls are made. This can be
+     * avoided by trampolining the call stack onto an {@link Executor}. For example:
+     *   {@code retryWhen((i, cause) -> i % 10 == 0 ? executor.submit(() -> { }) : Completable.completed())}
+     * </pre>
+     * This method provides a means to retry an operation under certain failure conditions in an asynchronous fashion
+     * and in sequential programming is similar to:
+     * <pre>{@code
+     *     public List<T> execute() {
+     *         List<T> results = ...;
+     *         return execute(0, results);
+     *     }
+     *
+     *     private List<T> execute(int attempts, List<T> results) {
+     *         try {
+     *             Iterator<T> itr = resultOfThisPublisher();
+     *             while (itr.hasNext()) {
+     *                 T t = itr.next(); // Any iteration with the Iterator may throw
+     *                 results.add(t);
+     *             }
+     *             return results;
+     *         } catch (Throwable cause) {
+     *             if (terminateOnNextException) {
+     *               throw cause;
+     *             }
+     *             try {
+     *                 shouldRetry.apply(attempts + 1, cause); // Either throws or completes normally
+     *                 execute(attempts + 1, results);
+     *             } catch (Throwable ignored) {
+     *                 throw cause;
+     *             }
+     *         }
+     *     }
+     * }</pre>
+     * @param terminateOnNextException
+     * <ul>
+     *     <li>{@code true} means that exceptions thrown from downstream {@link Subscriber#onNext(Object)} will be
+     *     caught, cancel the {@link Subscription}, propagate a {@link Subscriber#onError(Throwable)} downstream, and
+     *     no retry will be attempted. <b>Unless you are sure all downstream operators consume the
+     *     {@link Subscriber#onNext(Object)} this option SHOULD be used</b>. Otherwise, the outstanding demand counting
```

Comment:
my two cents, this is too much information for a ST user. This should be javadoc for the try...catch inside the operator. Consider keeping the first sentence only.

Reply:
I reduce the javadocs on these operators.